core/chains/evm/logpoller/log_poller_internal_test.go

```diff
@@ -380,14 +383,14 @@ func TestLogPoller_Replay(t *testing.T) {
 		}()
 		select {
 		case <-timeout:
-			assert.Fail(t, "lp.run() got stuck--failed to respond to second replay event within 1s")
+			assert.Fail(t, "lp.run() got stuck--failed to respond to second replay event within 2s")
```

Comment:
These pre-existing timeouts looks just as problematic, no?
`WaitGroupChan` sort of seems like a fundamental anti-pattern :thinking:

Reply:
Actually, I think this one is not a problem because it's `assert.Fail` rather than `require.Fail`?

If I understand the problem you've pointed out with the other changes correctly, the anti pattern should be about doing an early exit in response to finding there's a stuck goroutine somewhere.  As long as it just prints a FAIL message but continues to wait then using `WaitGroupChan` can't make anything worse than not using it.

I could improve this further by adding `lp.cancel()` here.  Without it, it will continue on to get stuck at the next `lp.wg.Wait()`, but with `lp.cancel()` it would have a good chance of getting unstuck so that it can go on to the next test.

Reply:
> As long as it just prints a FAIL message but continues to wait then using WaitGroupChan can't make anything worse than not using it.

I don't understand the relevant of assert vs. require here. Where does it continue to wait on `wg`? I only see the alternate case on 387.

Reply:
Line 393:
`lp.wg.Wait()`

Reply:
Oh, unless exiting early from a subtest has the same racey problem as exiting from a test?

Reply:
But `wg != lp.wg`. Does `lp.wg` completing necessarily mean that `wg` also completed?

Reply:
Ah, good point.  If everything's functioning properly one should imply the other, but we can't necessarily assume that I guess.

This `wg` is a bit different than the others in that it's not used to track goroutines.

Its purpose is to make sure that both calls to `FilterLogs()` happen within the specified time limit, at which point the test can be marked as PASS, otherwise it's a FAIL.  Its function is the same as an `assert.Eventually()` that waits for a condition to be true or times out, but without having to do any polling loop since the `On.Filter()` calls both signal it when they've been called.

The first call to FilterLogs() spawns a separate goroutine for sending the replay start signal.  That's already tested in other tests, so I was assuming that one will work regardless... and if that's true, then this is a non-issue.  However, it can be made more bulletproof by adding a non-blocking recv on replayStart to release that goroutine ands then calling `wg.Wait()` before allowing the subtest to complete.

I just went through and found several other places in these tests where the implicit assumption is that it's okay to exit a subtest even though there might be some goroutines still running if unexpected errors occur.   In particular, it seems we should always have these `lp.wg.Wait()` calls happen before each subtest exits rather than after.  I've moved them into `defer` blocks so that they should never allow a subtest to exit without all the threads being complete.  Also added some calls to `lp.cancel()`, `lp.Close()` etc. to help unblock things before before giving up and hanging the subtest.

Reply:
> This wg is a bit different than the others in that it's not used to track goroutines.

What do you mean by this? The mocks spawn goroutines that call `wg.Done()`, and one of them even calls `lp.Replay`, so it is necessarily possible for that method call to be in progress when this times out.

Reply:
Oh, sorry... just realized why what I wrote was confusing:  apparently, I was looking at the wrong subtest when I wrote that.

Both the "client abort doesnt hang run loop" subtest and the "shutdown during replay" subtest use a waitgroup to make sure both mocks have been called before the timeout is reached.  In the former, only one of the mocks spawns a goroutine whereas in the latter both do.

So in the latter case, which is what's actually above... you can think of it either way I guess.  It's either counting goroutines or counting mock calls, they happen to be the same.

I wasn't trying to say there was no problem with the waitgroup, just noting that it's being used a bit differently (at least the one I was looking at) from the typical use case.  Anyway, both of them call `wg.Wait()` now before the subtest can exit so, all these problems should be solved!